NextJS Codegen

seawatts opened the pull request:
This adds native integration for next.js code generation.

seawatts committed:
feat: generation for react working

feat: added nextjs-plugin package for quick next.js config setup

feat: add codelens for react

aaronvg commented on engine/language_client_codegen/baml_src/baml_client/__init__.py:
should this baml_client be here? in language_client_codegen?

seawatts replied:
Good catch, not sure how that got put in there.

aaronvg commented on integ-tests/react/src/app/api/baml/[baml_function]/route.ts:
```diff
@@ -0,0 +1,88 @@
+import { NextRequest, NextResponse } from 'next/server'
```
is this a file that users need to write / copy paste? Or does the generator try to put this route here?

seawatts replied:
Oops, we aren't using this. I forgot to delete it.

aaronvg commented on integ-tests/react/src/app/actions.ts:
```diff
@@ -0,0 +1,33 @@
+
```
is this also written by the generator? Outside of baml_client and directly into your app/ ?

seawatts replied:
Also not being used, deleting.

seawatts committed:
chore: delete unused files

seawatts committed:
chore: generate files

seawatts committed:
feat: sort types, classes, functions by name in code generation

imalsogreg commented on engine/language_client_codegen/src/openapi.rs:
```diff
@@ -326,22 +326,26 @@ impl<'ir> TryFrom<(&'ir IntermediateRepr, &'_ crate::GeneratorArgs)> for OpenApi
+                ))
+            })
+            .collect::<Result<_>>()?;
+        paths.sort_by(|a, b| a.function_name.cmp(b.function_name));
```
Thank you :)

seawatts replied:
🥳 

I was having an issue that I’m surprised we hadn’t seen before, where the open API and typescript generators were producing slightly different list ordering for types, functions, classes, etc. It was only intermittently, though. I added some updates to the rust generators to sort the lists by name beforehand so it will be consistent every time. I haven’t updated Python or Ruby yet, though.

imalsogreg commented on engine/language_client_codegen/src/typescript/mod.rs:
```diff
@@ -1,20 +1,82 @@
+            match output_type {
+                Some(GeneratorOutputType::TypescriptReact) => Self::React,
+                Some(GeneratorOutputType::Typescript) | None => Self::None,
+                Some(_) => panic!("Invalid generator type for TypeScript framework"),
```
Consider returning `Option<Self>`, with `None` for the wildcard case. If `Some(_)` is truly unreachable, the callsite is a clearer place to indicate that.

It's also better to explicitly match every possible `GeneratorOutputType`. Wildcard patterns (`_`) make it easy to accidentally add a variant, then forget to update the `match`. e.g. if we added `GeneratorOutputType::TypescriptSveltekit` to the `GeneratorOutputType`, `from_generator_type` would silently start to panic. But if we never use a wildcard pattern, we get a compiler error that points us to the function we need to update.

seawatts replied:
Fixed!

imalsogreg commented on engine/language_client_codegen/src/typescript/templates/types.ts.j2:
```diff
@@ -1,4 +1,8 @@
-import { Image } from "@boundaryml/baml"
+import type { Image, Audio } from "@boundaryml/baml"
+
+export type RecursivePartialNull<T> = T extends object
```
Do we still use/need `RecursivePartialNull`?

seawatts replied:
This is still being exported for people who might have been importing this in their projects. We are not using it anywhere else in the generated code. We could add a `@depreacted` flag. Thoughts?

seawatts replied:
Added one.

imalsogreg commented on engine/language_client_typescript/artifacts/async_context_vars.d.ts.map:
```diff
@@ -0,0 +1 @@
+{"version":3,"file":"async_context_vars.d.ts","sourceRoot":"","sources":["../typescript_src/async_context_vars.ts"],"names":[],"mappings":"AAAA,OAAO,EAAE,QAAQ,EAAE,qBAAqB,EAAE,WAAW,EAAE,YAAY,EAAE,MAAM,UAAU,CAAA;AAGrF,qBAAa,cAAc;IACzB,OAAO,CAAC,EAAE,CAAa;IACvB,OAAO,CAAC,GAAG,CAA0C;gBAEzC,EAAE,EAAE,WAAW;IAS3B,WAAW,IAAI,OAAO;IAYtB,KAAK,IAAI,IAAI;IAKb,UAAU,CAAC,IAAI,EAAE,MAAM,CAAC,MAAM,EAAE,MAAM,CAAC ...
```
Did we mean to check in this and `async_context_vars.d.ts`?  This file looks like it wast meant to be gitignored?

seawatts replied:
No, will delete

imalsogreg commented on engine/language_client_typescript/artifacts/native.d.ts:
```diff
@@ -0,0 +1,133 @@
+/* auto-generated by NAPI-RS */
```
Looks like lots of files under `artifacts` get checked in. Did we mean to exclude them from VCS?

seawatts replied:
Ah ha, yes we can delete these now. It gets auto generated during build time. This allows us to do debugging locally a bit easier

imalsogreg commented on engine/language_client_typescript/stream.js:
```diff
@@ -57,5 +57,28 @@ class BamlStream {
+     */
+    toStreamable() {
+        const stream = this;
+        return new ReadableStream({
```
Where does `ReadableStream` come from? Does this break users who don't have next or react installed?

seawatts replied:
That is built into Node, we don't need to import it from anywhere. It's available in the bowser environment and node environment. Not react specific

imalsogreg commented on fern/.cursorrules:
```diff
@@ -0,0 +1,320 @@
+You are the world's best documentation writer, renowned for your clarity, precision, and engaging style. Every piece of documentation you produce is:
```
Just making sure - we want to check this in?

seawatts replied:
I copied this from their fern llm.txt file. When we use cursor for files in that directory it can help cursor get the correct results. 

https://buildwithfern.com/learn/docs/developer-tools/llms-txt

Cursor just did an update where you can have multiple .cursorrules files in a repo and it takes precedence based on the file you talking about.

imalsogreg commented on fern/docs.yml:
```diff
@@ -83,10 +83,36 @@ navigation:
             icon: fa-brands fa-react
-            path: 01-guide/08-integrations/nextjs.mdx
+            contents:
+              # - page: Overview
```
Commented code probably deletable.

seawatts replied:
Done!

imalsogreg commented on integ-tests/README.md:
````diff
@@ -172,7 +172,13 @@ b. **Using .env File (Recommended for open source contributors)**
+      ```
+
+      <Note>
+      For local development, use `pnpm run build:debug` which is much faster. The full `pnpm run build` can take several minutes as it creates an optimized production build.
````
Thank you :)

imalsogreg reviewed (approved):
Left a couple comments, none of them critical.

Looks pretty great!

Note - I haven't run integ tests myself.

I did run `pnpm generate` but didn't carefully examine the results.

Happy to do the above things soon, just wanted to give my initial 👍 in case you are trying to merge quickly.

seawatts committed:
feat: update docs for next.js / react

feat: updated ref docs for python, ruby, openapi, and typescript

seawatts committed:
chore: remove generated artifacts files

seawatts committed:
fix: add warning for next.js turbo pack

seawatts committed:
fix: issue with panic on from_generator_type on tyepscript mod

seawatts committed:
chore: add deprecation warning to RecursivePartialNull

seawatts committed:
chore: run generate

sxlijin commented on engine/language_client_codegen/src/typescript/templates/partial_types.ts.j2:
```diff
@@ -1,6 +1,7 @@
-
-import * as types from "./types"
+import type { Image, Audio } from "@boundaryml/baml"
+import type { Checked, Check } from "./types"
+import type { {% for t in types %} {{ t }}{% if !loop.last %}, {% endif %}{% endfor %} } from "./types"
```
I think this is a name collision, that users can't use "Check" for their own type name - or is that a decision we already made? (Same also applies to "Image" and "Audio", but I think that one is less egregious.)

cc @imalsogreg

seawatts replied:
That's a good point. We have been doing that in other places it seems like. I'd say let's punt this to a new PR

sxlijin commented on engine/language_client_typescript/cli.js:
```diff
@@ -5,6 +5,6 @@ if (require.main === module) {
   }
 
-  const baml = require('./native')
+  const baml = require('./artifacts/native')
```
Can we split this change out of this PR?

I tried to do something along these lines a while back and screwed up the TS release in a hard-to-understand way. We can try another PR to push this forward but I'm uncomfortable doing that in this PR.

seawatts replied:
Okay, sounds good. I did it in order to do some testing locally where I need my integ-test/react project to reference the node wasm bindings and this is how I got it to work.

sxlijin commented on fern/01-guide/08-frameworks/01-react-nextjs/01-app-router.mdx:
````diff
@@ -0,0 +1,452 @@
+
+<CodeBlocks>
+```bash npm
+npm install @boundaryml/baml
+npm install @boundaryml/baml-nextjs-plugin
````
```suggestion
npm install @boundaryml/baml @boundaryml/baml-nextjs-plugin
```

and I think we can do the same for the others too?

seawatts replied:
Done

sxlijin reviewed (commented):
reading docs now

seawatts committed:
docs: update react/next ref and guide docs

seawatts committed:
chore: clean up artifacts and generate types

seawatts committed:
chore: update gitignore to not include artifcats

seawatts committed:
feat: update react docs

feat: change react generatd file structure
feat: change react output types
feat: update examples

seawatts committed:
fix: issues with ts and tests

seawatts committed:
Merge branch 'canary' into seawatts/next-js-codegen

seawatts committed:
fix: merge

seawatts committed:
chore: clean up docs for react

imalsogreg reviewed (approved):
🚀

seawatts committed:
chore: code gen

seawatts committed:
feat: change react hook input / output to final and stream

seawatts committed:
feat: update next docs as well as hook example in integ tests

seawatts committed:
Merge branch 'canary' into seawatts/next-js-codegen

seawatts committed:
chore: gen types

seawatts committed:
fix: type error
